### PR TITLE
Tackle files with a single line of coverage missing

### PR DESCRIPTION
**Pull request by jon-betts**
Based on:

 * https://github.com/hypothesis/h/pull/8085

## What this PR tries to fix

This PR focuses on fixing coverage issues in files where there is only one line of coverage missing. When I tested these comprised about 1/3 of the outstanding files with coverage issues.

### Why these files?

The idea here is these should tend to be the easiest to fix, while reducing the number of files with coverage issues the most. This means you are more likely to be able to spot the thing you are looking for if these could all be fixed.

Doing all test coverage in `h` at once is a big ask (both to do and review) so we need some ways to break the task up, which are essentially arbitrary. You could do each of these as a separate PR, but nobody wants to look at that.

### What I didn't manage to fix and why:

`h/schemas/forms/accounts/login.py`

This is code which is called from a `colander.deferred` widget. I have no idea what that is or how to test it.

`h/activity/query.py`

The tests for this code are pretty wild and do that thing where every line of code gets it's own test instead of running the code once and saying what it's supposed to do.

In the middle of this we fixture out a piece of code which reads groups from the DB. I don't know why we do this, but this is where the issues come from.

I'm not sure where to start with this one, I think you'd want to refactor this to be less first. I suspect we've ended up mocking out this part as a way of making assertions, because so much around this is mocked out too.

## Testing notes

There really shouldn't be any change in behavior here. There is barely any change to the code outside of the tests. I think all bar the date change there are kind of trivially fine. You can try the date TZ issue yourself with code like this:

```python
from dateutil.parser import parse

from h.search.query import DEFAULT_DATE

# Can have no TZ, defaults to the 31st
date = parse("2001-01")
print(date)  # 2001-01-31 00:00:00
print(date.tzinfo)  # None

# Has a TZ, has been defaulted to the first
date = parse("2001-01", default=DEFAULT_DATE)
print(date)  # 2001-01-01 00:00:00+00:00
print(date.tzinfo)  # tzutc()
```

For the tests:

 * `make backend-tests`
 * `make coverage | grep -v "-" | grep -v ","`
 * You should see:

```shell
...
h/activity/query.py                             83      1     32      0  99.13%   194
h/schemas/forms/accounts/login.py               44      1      6      1  96.00%   90
...
```

### CODE REVIEW

File: h/search/query.py [hunk 1]
```diff
@@ -142,10 +142,6 @@ def _parse_date(str_value):
         except ValueError:
             try:  # pylint: disable=too-many-try-statements
                 date = parse(str_value, default=DEFAULT_DATE)
-                # If timezone isn't specified assume it's utc.
-                if not date.tzinfo:
-                    date = date.replace(tzinfo=tz.tzutc())
-
```
Comment (jon-betts) [inline h/search/query.py:148]: There's a comment in the commit about this, but the DEFAULT_DATE provides values for every missing component, including TZ.

File: h/search/core.py [hunk 2]
```diff
@@ -157,5 +157,8 @@ def _parse_aggregation_results(self, aggregations):
     def _get_total_hits(response):
         total = response["hits"]["total"]
         if isinstance(total, int):
-            return total  # ES 6.x
-        return total["value"]  # ES 7.x
+            # ES 6.x
+            return total
+
+        # ES 7.x
+        return total["value"]  # pragma: nocover
```
Comment (jon-betts) [inline h/search/core.py:164]: This is work to extend our support for ES to 7, but we've not made the jump yet. When we do this coverage should break again and we can remove the 6 path.

File: tests/h/models/blocklist_test.py [hunk 3]
```diff
@@ -1,20 +1,26 @@
...
+            (None, "http://example.com", False),
+            ("%//example.com%", "http://example.com", True),
+            ("%//example.com%", "http://example.com/path", True),
+        ),
+    )
+    def test_is_blocked(self, db_session, block_uri, uri, is_blocked):
+        if block_uri:
+            db_session.add(Blocklist(uri=block_uri))
 
+        assert Blocklist.is_blocked(db_session, uri) == is_blocked
```
Comment (jon-betts) [inline tests/h/models/blocklist_test.py:21]: I tidied this up before adding the repr stuff, because it wasn't very nice

File: tests/h/services/document_test.py [hunk 4]
```diff
@@ -87,40 +89,46 @@ def test_it_returns_documents_ordered_by_last_activity_desc(
...
+        return DocumentService(session=db_session)
 
 
-@pytest.fixture
-def other_annotations(factories, groups):
-    return [
-        factories.Annotation(groupid=groups["other_group"].pubid, shared=True),
-        factories.Annotation(groupid=groups["other_group"].pubid, shared=True),
-        factories.Annotation(groupid=groups["other_group"].pubid, shared=True),
-    ]
```
Comment (jon-betts) [inline tests/h/services/document_test.py:121]: The changes above here are all indenting the methods to attach them to the relevant class instead of having them floating about at the bottom. This leaves space for our service factory test.

File: tests/h/services/document_test.py [hunk 5]
```diff
@@ -87,40 +89,46 @@ def test_it_returns_documents_ordered_by_last_activity_desc(
...
+class TestServiceFactory:
+    def test_it(self, pyramid_request, DocumentService):
+        svc = document_service_factory(sentinel.context, pyramid_request)
 
+        DocumentService.assert_called_once_with(session=pyramid_request.db)
+        assert svc == DocumentService.return_value
```
Comment (jon-betts) [inline tests/h/services/document_test.py:130]: In other products we call the factory method "factory" or "service_factory" rather than having a different name for each. The namespace does that for you.

I've named them assuming that will happen without doing the unrelated refactor of making it true.

File: tests/h/util/view_test.py [hunk 6]
```diff
@@ -5,27 +5,12 @@
...
-
-    @pytest.fixture
-    def latest_exception(self):
-        return Exception("Last exception raised in thread")
-
-    @pytest.fixture
-    def sys_exc_info(self, patch, latest_exception):
-        sys_exc_info = patch("h.util.view._exc_info")
-        sys_exc_info.return_value = (type(latest_exception), latest_exception, None)
-        return sys_exc_info
```
Comment (jon-betts) [inline tests/h/util/view_test.py:27]: ? I assume this made sense once?

File: tests/h/views/admin/documents_test.py [hunk 7]
```diff
@@ -53,6 +53,10 @@ def test_it_errors_if_validation_fails(
     def move_annotations_by_url(self, patch):
         return patch("h.views.admin.documents.move_annotations_by_url")
 
+    @pytest.fixture
+    def views(self, pyramid_request):
+        return DocumentsAdminViews(pyramid_request)
```
Comment (jon-betts) [inline tests/h/views/admin/documents_test.py:58]: Introduced this fixture as I wanted it in the get test above. Thought we might as well re-use it

File: h/services/document.py [hunk 8]
```diff
@@ -44,4 +44,4 @@ def fetch_by_groupid(self, groupid, userid=None):
 
 
 def document_service_factory(_context, request):
-    return DocumentService(request.db)
+    return DocumentService(session=request.db)
```
Comment (jon-betts) [inline h/services/document.py:47]: When I wrote the test for this I think it's nicer to use kwargs than positional args, as they are more refactoring friendly.

File: tests/h/models/activation_test.py [hunk 9]
```diff
@@ -1,12 +1,24 @@
...
+    def activation(self, db_session):
+        activation = Activation()
 
-    db_session.add(act)
-    db_session.flush()
+        db_session.add(activation)
+        db_session.flush()
 
-    assert re.match(r"[A-Za-z0-9]{12}", act.code)
+        return activation
```
Comment (jon-betts) [inline tests/h/models/activation_test.py:24]: I re-used most of the setup for the existing test as a fixture here.

File: tests/h/models/auth_client_test.py [hunk 10]
```diff
@@ -31,6 +31,11 @@ def test_allows_empty_redirect_uri_for_other_grant(
         db_session.add(client)
         db_session.flush()
 
+    def test___repr__(self):
+        client = AuthClient(id=123)
+
+        assert repr(client) == "AuthClient(id=123)"
```
Comment (jon-betts) [inline tests/h/models/auth_client_test.py:37]: I didn't try to make these consistent, just to make them pass coverage.

File: tests/h/models/blocklist_test.py [hunk 11]
```diff
@@ -1,20 +1,26 @@
-from h import models
+import pytest
 
+from h.models.blocklist import Blocklist
```
Comment (jon-betts) [inline tests/h/models/blocklist_test.py:3]: I like a 1:1 from the unit test to the file under test. We are here to test the `Blocklist` object, we can import it directly.

**Review by marcospri — APPROVED** [review]
98.59, almost there!
